File: flutter/lib/core/storage/note_repository.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:path_provider/path_provider.dart';
import 'package:path/path.dart' as path;
import '../crypto/key_manager.dart';
import 'vnc_format.dart';
import '../../domain/entities/note.dart';

/// Repository for note CRUD operations
/// Notes are stored as individual .vnc files in app's private storage
class NoteRepository {
  final KeyManager _keyManager;
  Directory? _notesDir;

  NoteRepository(this._keyManager);

  /// Initialize repository - get app directory
  Future<void> init() async {
    final appDir = await getApplicationDocumentsDirectory();
    _notesDir = Directory(path.join(appDir.path, 'notes'));
    if (!await _notesDir!.exists()) {
      await _notesDir!.create(recursive: true);
    }
  }

  /// Get file path for a note
  String _getNotePath(String noteId) {
    return path.join(_notesDir!.path, '$noteId.vnc');
  }

  /// Create a new note
  Future<void> createNote(Note note) async {
    if (!_keyManager.isUnlocked) {
      throw StateError('KeyManager not unlocked');
    }

    final data = await VNCFormat.encryptNote(
      note,
      _keyManager.masterKey!,
      _keyManager.hmacKey!,
    );

    final file = File(_getNotePath(note.id));
    await file.writeAsBytes(data);
  }

  /// Update an existing note
  Future<void> updateNote(Note note) async {
    await createNote(note); // Same as create (upsert)
  }

  /// Delete a note
  Future<void> deleteNote(String noteId) async {
    final file = File(_getNotePath(noteId));
    if (await file.exists()) {
      await file.delete();
    }
  }

  /// Get a single note by ID
  Future<Note?> getNote(String noteId) async {
    if (!_keyManager.isUnlocked) {
      throw StateError('KeyManager not unlocked');
    }

    final file = File(_getNotePath(noteId));
    if (!await file.exists()) {
      return null;
    }

    final data = await file.readAsBytes();
    return await VNCFormat.decryptNote(
      data,
      _keyManager.masterKey!,
      _keyManager.hmacKey!,
    );
  }

  /// Get all notes (including archived)
  Future<List<Note>> getAllNotes() async {
    if (!_keyManager.isUnlocked) {
      throw StateError('KeyManager not unlocked');
    }

    final notes = <Note>[];
    
    if (!await _notesDir!.exists()) {
      return notes;
    }

    final files = _notesDir!.listSync()
        .whereType<File>()
        .where((f) => f.path.endsWith('.vnc'));

    for (final file in files) {
      try {
        final data = await file.readAsBytes();
        final note = await VNCFormat.decryptNote(
          data,
          _keyManager.masterKey!,
          _keyManager.hmacKey!,
        );
        notes.add(note);
      } catch (e) {
        // Skip corrupted files
        print('Error reading note ${file.path}: $e');
      }
    }

    // Sort: pinned first, then by modified date descending
    notes.sort((a, b) {
      if (a.pinned && !b.pinned) return -1;
      if (!a.pinned && b.pinned) return 1;
      return b.modified.compareTo(a.modified);
    });

    return notes;
  }

  /// Get non-archived notes
  Future<List<Note>> getActiveNotes() async {
    final all = await getAllNotes();
    return all.where((n) => !n.archived).toList();
  }

  /// Get archived notes
  Future<List<Note>> getArchivedNotes() async {
    final all = await getAllNotes();
    return all.where((n) => n.archived).toList();
  }

  /// Search notes by text (decrypt and search in-memory)
  Future<List<Note>> searchNotes(String query) async {
    final all = await getActiveNotes();
    final lowerQuery = query.toLowerCase();
    
    return all.where((note) {
      return note.title.toLowerCase().contains(lowerQuery) ||
          note.body.toLowerCase().contains(lowerQuery) ||
          note.labels.any((l) => l.toLowerCase().contains(lowerQuery));
    }).toList();
  }

  /// Get notes by label
  Future<List<Note>> getNotesByLabel(String label) async {
    final all = await getActiveNotes();
    return all.where((n) => n.labels.contains(label)).toList();
  }

  /// Export a note as .vnc bytes (for sharing)
  Future<Uint8List> exportNote(String noteId) async {
    final note = await getNote(noteId);
    if (note == null) {
      throw Exception('Note not found');
    }

    return await VNCFormat.encryptNote(
      note,
      _keyManager.masterKey!,
      _keyManager.hmacKey!,
    );
  }

  /// Import a note from .vnc bytes
  Future<Note> importNote(Uint8List data) async {
    final note = await VNCFormat.decryptNote(
      data,
      _keyManager.masterKey!,
      _keyManager.hmacKey!,
    );
    
    // Save the imported note
    await createNote(note);
    return note;
  }
}

File: web/src/core/storage/noteRepository.ts
import { Note } from '../../domain/entities/types'
import { VNCFormat } from './vncFormat'

/**
 * Repository for note CRUD operations
 * Notes are stored in IndexedDB as .vnc encrypted files
 */
export class NoteRepository {
  private readonly dbName = 'VaultNote'
  private readonly storeName = 'notes'
  private db: IDBDatabase | null = null

  private key: Uint8Array | null = null
  private hmacKey: Uint8Array | null = null

  /**
   * Initialize the repository with encryption keys
   */
  setKeys(key: Uint8Array, hmacKey: Uint8Array): void {
    this.key = key
    this.hmacKey = hmacKey
  }

  /**
   * Initialize IndexedDB
   */
  async init(): Promise<void> {
    return new Promise((resolve, reject) => {
      const request = indexedDB.open(this.dbName, 1)

      request.onerror = () => reject(request.error)
      request.onsuccess = () => {
        this.db = request.result
        resolve()
      }

      request.onupgradeneeded = (event) => {
        const db = (event.target as IDBOpenDBRequest).result
        if (!db.objectStoreNames.contains(this.storeName)) {
          db.createObjectStore(this.storeName, { keyPath: 'id' })
        }
      }
    })
  }

  /**
   * Create a new note
   */
  async createNote(note: Note): Promise<void> {
    if (!this.key || !this.hmacKey) {
      throw new Error('Encryption keys not set')
    }

    const encryptedData = await VNCFormat.encryptNote(note, this.key, this.hmacKey)

    return new Promise((resolve, reject) => {
      if (!this.db) {
        reject(new Error('Database not initialized'))
        return
      }

      const transaction = this.db.transaction([this.storeName], 'readwrite')
      const store = transaction.objectStore(this.storeName)

      const request = store.put({
        id: note.id,
        data: encryptedData,
        modified: note.modified,
      })

      request.onsuccess = () => resolve()
      request.onerror = () => reject(request.error)
    })
  }

  /**
   * Update an existing note
   */
  async updateNote(note: Note): Promise<void> {
    await this.createNote(note)
  }

  /**
   * Delete a note
   */
  async deleteNote(noteId: string): Promise<void> {
    return new Promise((resolve, reject) => {
      if (!this.db) {
        reject(new Error('Database not initialized'))
        return
      }

      const transaction = this.db.transaction([this.storeName], 'readwrite')
      const store = transaction.objectStore(this.storeName)

      const request = store.delete(noteId)

      request.onsuccess = () => resolve()
      request.onerror = () => reject(request.error)
    })
  }

  /**
   * Get a single note by ID
   */
  async getNote(noteId: string): Promise<Note | null> {
    if (!this.key || !this.hmacKey) {
      throw new Error('Encryption keys not set')
    }

    return new Promise(async (resolve, reject) => {
      if (!this.db) {
        reject(new Error('Database not initialized'))
        return
      }

      const transaction = this.db.transaction([this.storeName], 'readonly')
      const store = transaction.objectStore(this.storeName)

      const request = store.get(noteId)

      request.onsuccess = async () => {
        if (!request.result) {
          resolve(null)
          return
        }

        try {
          const note = await VNCFormat.decryptNote(
            request.result.data,
            this.key!,
            this.hmacKey!
          )
          resolve(note)
        } catch (error) {
          reject(error)
        }
      }

      request.onerror = () => reject(request.error)
    })
  }

  /**
   * Get all notes
   */
  async getAllNotes(): Promise<Note[]> {
    if (!this.key || !this.hmacKey) {
      throw new Error('Encryption keys not set')
    }

    return new Promise(async (resolve, reject) => {
      if (!this.db) {
        reject(new Error('Database not initialized'))
        return
      }

      const transaction = this.db.transaction([this.storeName], 'readonly')
      const store = transaction.objectStore(this.storeName)

      const request = store.getAll()

      request.onsuccess = async () => {
        try {
          const notes: Note[] = []

          for (const record of request.result) {
            try {
              const note = await VNCFormat.decryptNote(
                record.data,
                this.key!,
                this.hmacKey!
              )
              notes.push(note)
            } catch (error) {
              console.error('Error decrypting note:', error)
            }
          }

          // Sort: pinned first, then by modified date descending
          notes.sort((a, b) => {
            if (a.pinned && !b.pinned) return -1
            if (!a.pinned && b.pinned) return 1
            return new Date(b.modified).getTime() - new Date(a.modified).getTime()
          })

          resolve(notes)
        } catch (error) {
          reject(error)
        }
      }

      request.onerror = () => reject(request.error)
    })
  }

  /**
   * Get non-archived notes
   */
  async getActiveNotes(): Promise<Note[]> {
    const all = await this.getAllNotes()
    return all.filter((n) => !n.archived)
  }

  /**
   * Get archived notes
   */
  async getArchivedNotes(): Promise<Note[]> {
    const all = await this.getAllNotes()
    return all.filter((n) => n.archived)
  }

  /**
   * Search notes by text
   */
  async searchNotes(query: string): Promise<Note[]> {
    const all = await this.getActiveNotes()
    const lowerQuery = query.toLowerCase()

    return all.filter(
      (note) =>
        note.title.toLowerCase().includes(lowerQuery) ||
        note.body.toLowerCase().includes(lowerQuery) ||
        note.labels.some((l) => l.toLowerCase().includes(lowerQuery))
    )
  }

  /**
   * Get notes by label
   */
  async getNotesByLabel(label: string): Promise<Note[]> {
    const all = await this.getActiveNotes()
    return all.filter((n) => n.labels.includes(label))
  }

  /**
   * Export a note as .vnc bytes
   */
  async exportNote(noteId: string): Promise<Uint8Array> {
    const note = await this.getNote(noteId)
    if (!note) {
      throw new Error('Note not found')
    }

    if (!this.key || !this.hmacKey) {
      throw new Error('Encryption keys not set')
    }

    return VNCFormat.encryptNote(note, this.key, this.hmacKey)
  }

  /**
   * Import a note from .vnc bytes
   */
  async importNote(data: Uint8Array): Promise<Note> {
    if (!this.key || !this.hmacKey) {
      throw new Error('Encryption keys not set')
    }

    const note = await VNCFormat.decryptNote(data, this.key, this.hmacKey)
    await this.createNote(note)
    return note
  }
}
